GeRM Updates for next cycle

jrmlhermitte opened the pull request:
GeRM updates for next cycle.

- Added/removed processing functions (in handlers.py, calibration.py, reduction.py, stitching.py)
- Added docs which I temporarily built elsewhere
- Tested on some GeRM data locally and it works (see the docs link for the images of the examples)

See the [docs](https://pygerm.readthedocs.io/en/make-histogram-docs/) for all details.

tacaswell commented on pygerm/reduction.py:
```diff
@@ -1,144 +1,297 @@
+            (I was told it's on the order of 12 clock cyles, much less
+            less than my aggressive threhsold of 2**26)
+    '''
+    # if it's dask array compute it
```
Would `ts = np.array(ts)` work here?

jrmlhermitte replied:
this works thanks! I think we should put `copy=True` just in case (it's `True` by default currently though)

tacaswell commented on pygerm/reduction.py:
```diff
@@ -1,144 +1,297 @@
+                - the times (in s)
+
+    '''
+    if hasattr(germ_ts, 'compute'):
```
There has to be a way to do this with out resorting to `hasattr` checks?  I think `np.asarray` will be lazy for numpy arrays and force comptuation of dask arrays (or anything else that respects the `__array__` protocol).

jrmlhermitte replied:
this is perfect and simple, tested and worked thanks!

tacaswell commented on pygerm/reduction_old.py:
```diff
@@ -0,0 +1,143 @@
+import pandas as pd
```
if we don't need this just delete it.  If we do need this, keep it where it was please.

jrmlhermitte replied:
ok i'll delete most, ive written comments as a review as to why im removing/keeping stuff there.

jrmlhermitte commented on pygerm/reduction_old.py:
```diff
@@ -0,0 +1,143 @@
+import matplotlib.pyplot as plt
+
+
+def bin_frame(data, bins, corr_mat=None):
```
Replaced

this is replaced with the more general `histogram_germ` (using `skbeam.core.accumulators.histogram.Histogram` under the hood)
(keeping notes)

jrmlhermitte commented on pygerm/reduction_old.py:
```diff
@@ -0,0 +1,143 @@
+    return spectrum, bin_edges
+
+
+def select_energy_band(spectrum, energy_bins, lo, hi):
```
Maybe keep.

this can also be done when histogramming the bins in the first step (you can specify energy)

jrmlhermitte commented on pygerm/reduction_old.py:
```diff
@@ -0,0 +1,143 @@
+    return spectrum[lo_ind:hi_ind].sum(axis=0)
+
+
+def stack_1D_spectrum(spectrum_list, energy_bins, lo, hi):
```
keep
this would still be useful

jrmlhermitte commented on pygerm/reduction_old.py:
```diff
@@ -0,0 +1,143 @@
+                     for s in spectrum_list])
+
+
+def sum_diffraction(diff_list, detector_angles, angle_bins, *,
```
replaced
this is replaced by `stitch`

jrmlhermitte commented on pygerm/reduction_old.py:
```diff
@@ -0,0 +1,143 @@
+                       (np.arange(384) - 192) * pixel_scale +
+                       pixel_offset)
+
+        out += np.histogram(line_angles, weights=line, bins=angle_bins)[0]
```
You should interpolate for the weights, when a shift is a fraction of a pixel. `stitch` does this by using interpolation on the weights (crude, but works)
Also, I made `stitch` work in a streamable fashion: i.e. you don't need to specify angle bins, just the delta angle (assumed to be the same per step)
`stitch` will expand the array as needed. So I think we can also remove this.

This is a great routine though! We could keep is as `stich2_alternative` or something as well.

jrmlhermitte commented on pygerm/reduction_old.py:
```diff
@@ -0,0 +1,143 @@
+    ax2.imshow(filter_peaks, aspect='auto', origin='lower',
+               extent=(-.5, 383.5, angles[0], angles[1]))
+
+    for _, angle, dist in zip(*hough_line_peaks(h, theta, d)):
```
for records, `hough_line_peaks` is a `skimage` routine [here](http://scikit-image.org/docs/dev/api/skimage.transform.html#skimage.transform.hough_line_peaks).

jrmlhermitte commented on pygerm/reduction_old.py:
```diff
@@ -0,0 +1,143 @@
+    return out / norm, angle_bins
+
+
+def calibrate_detector(peaks, angles, thresh=None):
```
We could never get the Hough transform to work properly so I took a different approach.
I fit per line and obtained a delta step.
This strongly depends on a calibration scan that is run on something with obvious peaks. It also helps to have the shifts be less than 200 channels or so (but this could potentially be run over much more)

Algorithm
For each line[i-1], line[i]:
cross correlatd line[i-1] and line[i]
find peak
save that peak position, that's the shift

I do this in a vectorized fashion (I compute all cross correlations at once)

I'll write more later but these are records to justify why I have chosen to remove this. Please do let me know if you disagree.

tacaswell commented on docs/source/handlers.rst:
```diff
@@ -0,0 +1,119 @@
+
+
+
+Connecting with databroker
```
This should be the first section, the first section should be regulated to "advanced debugging"

tacaswell commented:
Can you please put the comments about removal in the commit messages?

jrmlhermitte commented:
modified the commit [here](https://github.com/NSLS-II/GeRM/pull/23/commits/147cfdd3c8912dafb742b5f887de606ab1105050) thanks!

jrmlhermitte commented:
The code is ready to submit:
- Added many routines for calibration, histogramming and stitching
- Built the documentation from my own branch [here](https://pygerm.readthedocs.io/en/make-histogram-docs/) for now. (was a bit of a pain, readthedocs is py3.5!!!! had to comment out fstrings, in  separate branch)

Does this look fine?

For the run, what's in this documentation is what we should run.
For the stitching, it's actually streaming capable, with the way it's written so we could give streaming a shot.

I don't think it's worth spending more time until we start measurements with how much time we have to spend on other tasks.

What do you think? I apologize this is a long PR but it was simpler to add the docs to the same PR than make two PR's and have to rebase.

danielballan commented:
You'll need to install sphinx and doctr and any sphinx extensions you use. I suggest renaming `test-requirements.txt` to `requirements-dev.txt` and adding them there.

jrmlhermitte commented:
oops yes, i had a requirements-dev (from the cookiecutter thanks!) but I forgot to point to it. I just updated .travis.yml

I also moved doctr there. Technically, we don't need it when we're not building the docs (so could be in the if statement but it's probably cleaner to keep it in the requirements file)
